servicetalk-opentelemetry-http/src/main/java/io/servicetalk/opentelemetry/http/ServicetalkHttpClientAttributesGetter.java

```diff
@@ -0,0 +1,82 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicetalk.opentelemetry.http;
+
+import io.servicetalk.http.api.HttpRequestMetaData;
+import io.servicetalk.http.api.HttpResponseMetaData;
+import io.servicetalk.transport.api.HostAndPort;
+
+import io.opentelemetry.instrumentation.api.instrumenter.http.HttpClientAttributesGetter;
+
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+final class ServicetalkHttpClientAttributesGetter
+    implements HttpClientAttributesGetter<HttpRequestMetaData, HttpResponseMetaData> {
+
+    static final ServicetalkHttpClientAttributesGetter INSTANCE =
+        new ServicetalkHttpClientAttributesGetter();
+
+    private ServicetalkHttpClientAttributesGetter() {
+    }
+
+    @Override
+    public String getHttpRequestMethod(HttpRequestMetaData httpRequestMetaData) {
+        return httpRequestMetaData.method().name();
+    }
+
+    @Override
+    public List<String> getHttpRequestHeader(HttpRequestMetaData httpRequestMetaData, String name) {
+        CharSequence value = httpRequestMetaData.headers().get(name);
+        if (value != null) {
+            return Collections.singletonList(value.toString());
+        }
+        return Collections.emptyList();
+    }
+
+    @Override
+    public Integer getHttpResponseStatusCode(HttpRequestMetaData httpRequestMetaData,
+                                             HttpResponseMetaData httpResponseMetaData,
+                                             @Nullable Throwable error) {
+        return httpResponseMetaData.status().code();
+    }
+
+    @Override
+    public List<String> getHttpResponseHeader(HttpRequestMetaData httpRequestMetaData,
+                                              HttpResponseMetaData httpResponseMetaData,
+                                              String name) {
+        CharSequence value = httpResponseMetaData.headers().get(name);
+        if (value != null) {
+            return Collections.singletonList(value.toString());
+        }
+        return Collections.emptyList();
+    }
+
+    @Nullable
+    @Override
+    public String getUrlFull(HttpRequestMetaData request) {
+        HostAndPort effectiveHostAndPort = request.effectiveHostAndPort();
+        String requestScheme = request.scheme() != null ? request.scheme()
+            : "http";
+        CharSequence hostAndPort = request.headers().contains("host") ? request.headers().get("host")
+            : effectiveHostAndPort != null ?
+            String.format("%s:%d", effectiveHostAndPort.hostName(), effectiveHostAndPort.port()) : null;
+        return hostAndPort != null ? String.format("%s://%s%s", requestScheme, hostAndPort, request.path()) :
```

Comment:
The whole method has multiple issues:

1. `request.scheme()` is always `null` today for single-address client. Falling back to `http` is not correct. Please open a GH issue, and we can try to prefill scheme based on `SslConfig` configured at the client builder.
2. There is no need to check `host` header if you already parsed `effectiveHostAndPort`. Look at its javadoc, it already takes the `host` header into account. In cases when you need a header, use constants instead of hard coded header names.
3. Usage of `String.format` is strongly discouraged, it has the worst performance. Consider `+` or `StringBuilder` as alternatives.
4. `request.path()` is not enough, javadoc for this method described that you need everything: path, query parameter, fragment. 

For now, consider only returning `request.requestTarget()`. It will contain everything you need and will give a full URL if users passed it in absolute-form. Then we can improve after (1) is resolved.
In addition, consider implementing `UrlAttributesExtractor`, it will give users path, query, etc.

Reply:
ok, I refactored this a little.
I created the issue https://github.com/apple/servicetalk/issues/2674